src/datatype.c

```diff
@@ -140,3 +159,798 @@ pg_tle_create_shell_type_if_not_exists(PG_FUNCTION_ARGS)
+	char	   *typeName = NameStr(*PG_GETARG_NAME(1));
+	Oid			inputFuncId = PG_GETARG_OID(2);
+	Oid			outputFuncId = PG_GETARG_OID(3);
+	int16		internalLength = PG_GETARG_INT16(4);
+	char	   *probin = get_probin(fcinfo->flinfo->fn_oid);
+
+	create_base_type(typeNamespace, typeName, inputFuncId, outputFuncId, internalLength, probin, false);
+
+	/*
+	 * Make effects of commands visible
+	 */
+	CommandCounterIncrement();
+
+	PG_RETURN_VOID();
+}
+
+/*
+ * Registers a new base type if not exists (otherwise do nothing), fails if the base type cannot be defined.
+ *
+ */
+PG_FUNCTION_INFO_V1(pg_tle_create_base_type_if_not_exists);
+Datum
+pg_tle_create_base_type_if_not_exists(PG_FUNCTION_ARGS)
+{
+	Oid			typeNamespace = PG_GETARG_OID(0);
+	char	   *typeName = NameStr(*PG_GETARG_NAME(1));
+	Oid			inputFuncId = PG_GETARG_OID(2);
+	Oid			outputFuncId = PG_GETARG_OID(3);
+	int16		internalLength = PG_GETARG_INT16(4);
+	char	   *probin = get_probin(fcinfo->flinfo->fn_oid);
+	bool		result;
+
+	result = create_base_type(typeNamespace, typeName, inputFuncId, outputFuncId, internalLength, probin, true);
+
+	/*
+	 * Make effects of commands visible
+	 */
+	CommandCounterIncrement();
+
+	PG_RETURN_BOOL(result);
+}
+
+/*
+ * create_base_type
+ *
+ * Create a new base type, returns true when a new base type is successfully created.
+ *
+ * if_not_exists: if true, don't fail on duplicate name, just print a notice and return false.
+ * Otherwise, fail on duplicate name.
+ */
+static bool
+create_base_type(Oid typeNamespace, char *typeName, Oid inputFuncId, Oid outputFuncId, int16 internalLength, char *funcProbin, bool if_not_exists)
+{
+	AclResult	aclresult;
+	Oid			inputOid;
+	Oid			outputOid;
+	Oid			typeOid;
+	Oid			array_oid;
+	char	   *array_type;
+	ObjectAddress address;
+	Oid			inputFuncParamType;
+	Oid			outputFuncParamType;
+	char	   *namespaceName;
+
+	/*
+	 * Even though the SQL function is locked down so only a member of
+	 * pgtle_admin can run this function, let's check and make sure there is
+	 * not a way to bypass that
+	 */
+	check_is_pgtle_admin();
+
+	if (!(internalLength > 0 || internalLength == -1))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
+				 errmsg("invalid type internal length %d", internalLength)));
+
+	if (internalLength > 0)
+		internalLength += VARHDRSZ;
+
+	/* Check we have creation rights in target namespace */
+	aclresult = PG_NAMESPACE_ACLCHECK(typeNamespace, GetUserId(), ACL_CREATE);
+	namespaceName = get_namespace_name(typeNamespace);
+	if (aclresult != ACLCHECK_OK)
+		aclcheck_error(aclresult, OBJECT_SCHEMA, namespaceName);
+
+	/*
+	 * Look to see if type already exists
+	 */
+#if PG_VERSION_NUM >= 120000
+	typeOid = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid,
+							  CStringGetDatum(typeName),
+							  ObjectIdGetDatum(typeNamespace));
+#else
+	typeOid = GetSysCacheOid2(TYPENAMENSP,
+							  CStringGetDatum(typeName),
+							  ObjectIdGetDatum(typeNamespace));
+#endif
```

Comment:
This would be easier to read as a macro in compatibility.h:

```c
#if PG_VERSION >= 120000
#define GETSYSCACHEOID2(a,b,c)  GetSysCacheOid2(a, Anum_pg_type_oid, b, c);
#else
#define GETSYSCACHEOID2(a,b,c)  GetSysCacheOid2(a, b, c);
#endif
```

Then the code can become:

```suggestion
	typeOid = GETSYSCACHEOID2(TYPENAMENSP,
							  CStringGetDatum(typeName),
							  ObjectIdGetDatum(typeNamespace));
```

Reply:
Done. Thanks for the suggestion.